### PR TITLE
[FLINK-26028][Connector/pulsar] add sink documentation; change some source documentation

### PR DESCRIPTION
**Pull request by imaffe**
…ulsar source documentation.

## What is the purpose of the change

Add Chinese and English version of Pulsar Sink.

## Brief change log

- add pulsar sink to the auto doc generation configuration.
- generate 2 config options file

## Documentation

Yes

### CODE REVIEW

**Comment by syhily** [comment]
@fapaul The document is also ready. Can you check it?

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 1]
```diff
@@ -96,64 +101,65 @@ Pulsar source provide two ways of topic-partition subscription:
...
+You can use `setTopicPattern("topic-*", RegexSubscriptionMode.AllTopics)` to consume
+both `persistent` and `non-persistent` topics based on the topic pattern.
+Pulsar source would filter the available topics by the `RegexSubscriptionMode`.
 
 ### Deserializer
 
-A deserializer (Deserialization schema) is required for parsing Pulsar messages. The deserializer is
-configured by `setDeserializationSchema(PulsarDeserializationSchema)`.
+A deserializer (`PulsarDeserializationSchema`) is for parsing Pulsar messages from bytes.
+You can configure the deserializer using `setDeserializationSchema(PulsarDeserializationSchema)`.
```
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/pulsar.md:158]:
```suggestion
A deserializer (`PulsarDeserializationSchema`) is for decoding Pulsar messages from bytes.
You can configure the deserializer using `setDeserializationSchema(PulsarDeserializationSchema)`.
```

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 2]
```diff
@@ -287,65 +292,67 @@ Built-in stop cursors include:
...
 
 ### Dynamic Partition Discovery
 
 To handle scenarios like topic scaling-out or topic creation without restarting the Flink
-job, Pulsar source can be configured to periodically discover new partitions under provided
-topic-partition subscribing pattern. To enable partition discovery, set a non-negative value for
-the option `PulsarSourceOptions.PULSAR_PARTITION_DISCOVERY_INTERVAL_MS`:
+job, Pulsar source periodically discover new partitions under a provided
+topic-partition subscription pattern. To enable partition discovery, you can set a non-negative value for
+the `PulsarSourceOptions.PULSAR_PARTITION_DISCOVERY_INTERVAL_MS` option:
```
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/pulsar.md:338]:
```suggestion
job, the Pulsar source periodically discover new partitions under a provided
topic-partition subscription pattern. To enable partition discovery, you can set a non-negative value for
the `PulsarSourceOptions.PULSAR_PARTITION_DISCOVERY_INTERVAL_MS` option:
```

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 3]
```diff
@@ -387,27 +394,404 @@ You should enable transaction in the Pulsar `borker.conf` file when using these
...
+
+Pulsar Sink supports writing records into one or more Pulsar topics or a specified list of Pulsar partitions.
+
+{{< hint info >}}
+This part describes the Pulsar sink based on the new
+[data sink](https://cwiki.apache.org/confluence/display/FLINK/FLIP-191%3A+Extend+unified+Sink+interface+to+support+small+file+compaction) API.
+
+If you still want to use the legacy `SinkFunction` or on Flink 1.14 or previous releases, just use the StreamNative's
+[pulsar-flink](https://github.com/streamnative/pulsar-flink).
+{{< /hint >}}
```
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/pulsar.md:421]: Like for the Source, just remove parts that are referring to older versions

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 4]
```diff
@@ -387,27 +394,404 @@ You should enable transaction in the Pulsar `borker.conf` file when using these
...
+
+It is **recommended** to set the producer name in Pulsar Sink by `setProducerName(String)`.
+This gives a unique name to the flink connector in the Pulsar statistic dashboard.
+Give you a mechanism to monitor the performance of your flink applications.
+
+### Writing Topics
+
+Setting the topics to write is a bit like the [topic-partition subscription](#topic-partition-subscription)
+in Pulsar source. We support a mixin style of topic setting. Therefore, you can provide a list of topics,
+partitions, or both of them.
```
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/pulsar.md:457]:
```suggestion
Defining the topics for producing is similar to the [topic-partition subscription](#topic-partition-subscription)
in the Pulsar source. We support a mix-in style of topic setting. You can provide a list of topics,
partitions, or both of them.
```

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 5]
```diff
@@ -387,27 +394,404 @@ You should enable transaction in the Pulsar `borker.conf` file when using these
...
+
+Configure writing targets can be replaced by using a custom [`TopicRouter`]
+[message routing](#message-routing). And read [flexible topic naming](#flexible-topic-naming)
+for understanding how to configure partitions on the Pulsar connector.
+
+{{< hint warning >}}
+If you build  the Pulsar sink based on both the topic and its corresponding partitions, Pulsar sink merges them and only use the topic.
+
+For example, it uses the `PulsarSink.builder().setTopics("some-topic1", "some-topic1-partition-0")` option to build the Pulsar sink,
+it is simplified to `PulsarSink.builder().setTopics("some-topic1")`.
```
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/pulsar.md:481]:
```suggestion
For example, when using the `PulsarSink.builder().setTopics("some-topic1", "some-topic1-partition-0")` option to build the Pulsar sink,
this is simplified to `PulsarSink.builder().setTopics("some-topic1")`.
```

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 6]
```diff
@@ -387,27 +394,404 @@ You should enable transaction in the Pulsar `borker.conf` file when using these
...
+
+### Serializer
+
+A serializer (`PulsarSerializationSchema`) is required to serializing the record instance into bytes.
+Similar to `PulsarSource`, Pulsar sink supports both Flink's `SerializationSchema` and
+Pulsar's `Schema`. But Pulsar's `Schema.AUTO_PRODUCE_BYTES()` is not supported in Pulsar Sink.
+
+If you do not need the message key and other message properties in Pulsar's
+[Message](https://pulsar.apache.org/api/client/2.9.0-SNAPSHOT/org/apache/pulsar/client/api/Message.html) interface,
+you can use the predefined `PulsarSerializationSchema`. Pulsar sink provides two implementation methods.
```
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/pulsar.md:492]:
```suggestion
you can use the predefined `PulsarSerializationSchema`. The Pulsar sink provides two implementation methods.
```

**Review by MartijnVisser — CHANGES_REQUESTED** [review]
Thanks @imaffe for the PR - I haven't been able to fully review it yet, but I already wanted to give you some feedback. For future improvements, it would have been nice to have separate commits. One for the changes to the Source documentation, one for the new Sink documentation etc. I'll try to completely finish the review tomorrow.

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 7]
```diff
@@ -387,27 +394,404 @@ You should enable transaction in the Pulsar `borker.conf` file when using these
...
+{{< /hint >}}
+
+### Serializer
+
+A serializer (`PulsarSerializationSchema`) is required to serializing the record instance into bytes.
+Similar to `PulsarSource`, Pulsar sink supports both Flink's `SerializationSchema` and
+Pulsar's `Schema`. But Pulsar's `Schema.AUTO_PRODUCE_BYTES()` is not supported in Pulsar Sink.
+
+If you do not need the message key and other message properties in Pulsar's
+[Message](https://pulsar.apache.org/api/client/2.9.0-SNAPSHOT/org/apache/pulsar/client/api/Message.html) interface,
```
[the remark this answers is not in the document]
Comment (imaffe) [reply docs/content/docs/connectors/datastream/pulsar.md:491]: This is my mistake though this refers to a concept that is pretty stable. I'l; change it to use the latest stable version.

File: docs/content/docs/connectors/datastream/pulsar.md (hunk 3 above)
Comment (imaffe) [reply docs/content/docs/connectors/datastream/pulsar.md:418]: Sure, I'll remove it in the final commit ~

**Comment by imaffe** [comment]
I noticed that I
- misuse of `The`
- used lot of `would`
- like to use passive voice
- like to use long sentences and clauses
- etc..

I'll avoid these next time :) Thank you so much @MartijnVisser for this detailed review ~

File: docs/content/docs/connectors/datastream/pulsar.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (syhily) [reply docs/content/docs/connectors/datastream/pulsar.md:487]: Pulsar use `SNAPSHOT` javadocs. There is no stable version. I'm afraid we have to use the snapshot here.

**Comment by imaffe** [comment]
@MartijnVisser Hi Martin, I've updated some requested changes~

**Comment by MartijnVisser** [comment]
@imaffe I'm currently sick but I'll try to have a look asap

**Comment by imaffe** [comment]
> @imaffe I'm currently sick but I'll try to have a look asap

Take care Martin~ Health always comes first~

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 8]
```diff
@@ -387,27 +391,402 @@ You should enable transaction in the Pulsar `borker.conf` file when using these
...
+  It means that this mode has the highest throughput.
+- `AT_LEAST_ONCE`: No data loss happens, but data duplication can happen after a restart from checkpoint.
+- `EXACTLY_ONCE`: No data loss happens. Each record is sent to the Pulsar broker only once.
+  Pulsar Sink uses [Pulsar transaction](https://pulsar.apache.org/docs/en/transactions/)
+  and two-phase commit (2PC) to ensure records are sent only once even after pipeline restarts.
+
+### Delayed message delivery
+
+[Delayed message delivery](https://pulsar.apache.org/docs/en/next/concepts-messaging/#delayed-message-delivery)
+enables you to consume a message later. With delayed message enabled,  the Pulsar sink sends a message to the Pulsar topic
```
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/pulsar.md:594]:
```suggestion
enables you to delay the possibility to consume a message. With delayed message enabled, the Pulsar sink sends a message to the Pulsar topic
```

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 9]
```diff
@@ -387,27 +391,402 @@ You should enable transaction in the Pulsar `borker.conf` file when using these
...
+**immediately**, but the message is delivered to a consumer once the specified delay is over.
+
+Delayed message delivery only works in the `Shared` subscription type. In `Exclusive` and `Failover`
+subscription types, the delayed message is dispatched immediately.
+
+You can configure the `MessageDelayer` for judging when sending the message to the consumer.
```
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/pulsar.md:600]:
```suggestion
You can configure the `MessageDelayer` to define when to send the message to the consumer.
```

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 10]
```diff
@@ -387,27 +391,402 @@ You should enable transaction in the Pulsar `borker.conf` file when using these
...
+The default delayer is never delay the message dispatching. You can use the `MessageDelayer.fixed(Duration)` option to
```
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/pulsar.md:601]:
```suggestion
The default option is to never delay the message dispatching. You can use the `MessageDelayer.fixed(Duration)` option to
```

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 11]
```diff
@@ -387,27 +391,402 @@ You should enable transaction in the Pulsar `borker.conf` file when using these
...
+- `numBytesOut`, `numRecordsOut`, `numRecordsOutErrors` are retrieved from Pulsar client metrics.
+
+- `currentSendTime` tracks the time from when the producer calls `sendAync()` to
+  the time when the message is acknowledged by the broker. This metric is not available in `NONE` delivery guarantee.
+{{< /hint >}}
+
+The Pulsar producer refreshes its stats every 60 seconds by default. And PulsarSink retrieves the Pulsar producer
+stats every 500ms. Thus `numRecordsOut`, `numBytesOut`, `numAcksReceived`, and `numRecordsOutErrors` 
+is updated every 60 seconds. To increase the metrics refresh frequency, you can change
+the Pulsar producer stats refresh interval to a smaller value (minimum 1 second), as shown below.
```
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/pulsar.md:741]:
```suggestion
The Pulsar producer refreshes its stats every 60 seconds by default. The PulsarSink retrieves the Pulsar producer
stats every 500ms. That means that `numRecordsOut`, `numBytesOut`, `numAcksReceived`, and `numRecordsOutErrors` 
are updated every 60 seconds. To increase the metrics refresh frequency, you can change
the Pulsar producer stats refresh interval to a smaller value (minimum 1 second), as shown below.
```

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 12]
```diff
@@ -387,27 +391,402 @@ You should enable transaction in the Pulsar `borker.conf` file when using these
...
+builder.setConfig(PulsarOptions.PULSAR_STATS_INTERVAL_SECONDS. 1L)
+```
+
+`numBytesOutRate` and `numRecordsOutRate` are calculated based on the `numBytesOut` and `numRecordsOUt`
+counter respectively. Flink internally uses a fixed 60 seconds window to calculate the rates.
+
+### Brief Design Rationale
+
+Pulsar sink follow the Sink API defined in 
+[FLIP-191](https://cwiki.apache.org/confluence/display/FLINK/FLIP-191%3A+Extend+unified+Sink+interface+to+support+small+file+compaction).
```
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/pulsar.md:753]: I think you mean FLIP-143? https://cwiki.apache.org/confluence/display/FLINK/FLIP-143%3A+Unified+Sink+API

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 13]
```diff
@@ -387,27 +391,402 @@ You should enable transaction in the Pulsar `borker.conf` file when using these
...
+
+#### Stateless SinkWriter
+
+In `EXACTLY_ONCE` mode, the Pulsar sink does not store transaction information in checkpoint,
+which means a new bunch of transactions will be created after a restart.
+Therefore, any message in previous pending transactions is  either aborted or timeout
+(In no way they are visible to the downstream Pulsar consumer).
+Pulsar team is working to optimize the resource cost bought by unfinished pending transactions.
```
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/pulsar.md:761]:
```suggestion
In `EXACTLY_ONCE` mode, the Pulsar sink does not store transaction information in a checkpoint.
That means that new transactions will be created after a restart.
Therefore, any message in previous pending transactions is either aborted or timed out
(They are never visible to the downstream Pulsar consumer).
The Pulsar team is working to optimize the needed resources by unfinished pending transactions.
```

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 14]
```diff
@@ -387,27 +391,402 @@ You should enable transaction in the Pulsar `borker.conf` file when using these
...
+
+## Known Issues
+
+This section describes some known issues about the Pulsar connectors.
+
+### No TransactionCoordinatorNotFound, but automatic reconnect
+
+Pulsar  transaction is still in active development and is not stable. Pulsar 2.9.2
+introduce [a break change](https://github.com/apache/pulsar/pull/13135) in transactions.
+If you use Pulsar 2.9.2 or higher version  with an older Pulsar client, you might get a `TransactionCoordinatorNotFound` exception.
```
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/pulsar.md:778]:
```suggestion
Pulsar transactions are still in active development and are not stable. Pulsar 2.9.2
introduces [a break change](https://github.com/apache/pulsar/pull/13135) in transactions.
If you use Pulsar 2.9.2 or higher with an older Pulsar client, you might get a `TransactionCoordinatorNotFound` exception.
```

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 15]
```diff
@@ -387,27 +391,402 @@ You should enable transaction in the Pulsar `borker.conf` file when using these
...
+#### Pulsar Schema Evolution
+
+[Pulsar Schema Evolution](https://pulsar.apache.org/docs/en/schema-evolution-compatibility/) allows
+you to reuse the same Flink job after certain "allowed" data model changes, like adding or deleting
+a field in a AVRO-based Pojo class. Please note that you can specify Pulsar schema validation rules
+and define an auto schema update. For details, refer to [Pulsar Schema Evolution](https://pulsar.apache.org/docs/en/schema-evolution-compatibility/).
+
+## Known Issues
+
+This section describes some known issues about the Pulsar connectors.
```
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/pulsar.md:772]: Don't we also still have a known issue with Java 11 support? I would like to have that listed too

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 16]
```diff
@@ -387,27 +391,402 @@ You should enable transaction in the Pulsar `borker.conf` file when using these
...
+
 ## Upgrading to the Latest Connector Version
 
 The generic upgrade steps are outlined in [upgrading jobs and Flink versions guide]({{< ref "docs/ops/upgrading" >}}).
 The Pulsar connector does not store any state on the Flink side. The Pulsar connector pushes and stores all the states on the Pulsar side.
 For Pulsar, you additionally need to know these limitations:
 
 * Do not upgrade the Pulsar connector and Pulsar broker version at the same time.
-* Always use a newer Pulsar client with Pulsar connector for consuming message from Pulsar.
+* Always use a newer Pulsar client with Pulsar connector to consuming messages from Pulsar.
```
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/pulsar.md:789]:
```suggestion
* Always use a newer Pulsar client with Pulsar connector to consume messages from Pulsar.
```

**Review by MartijnVisser — CHANGES_REQUESTED** [review]
@imaffe Thanks for the updates, this should be good to go to merge in the next round. Please also check if you can squash your commits and rebase.

File: docs/content/docs/connectors/datastream/pulsar.md (hunk 12 above)
Comment (imaffe) [reply docs/content/docs/connectors/datastream/pulsar.md:753]: Well the PulsarSInk is based on SinkV2 which is the FLIP-191 mentioned API

File: docs/content/docs/connectors/datastream/pulsar.md [hunk 17]
```diff
@@ -44,17 +43,19 @@ See how to link with them for cluster execution [here]({{< ref "docs/dev/configu
 This part describes the Pulsar source based on the new
 [data source]({{< ref "docs/dev/datastream/sources.md" >}}) API.
 
-If you want to use the legacy `SourceFunction` or on Flink 1.13 or lower releases, just use the StreamNative's [pulsar-flink](https://github.com/streamnative/pulsar-flink).
+If you want to use the legacy `SourceFunction` or on Flink 1.13 or lower releases,
+just use the StreamNative's [Pulsar Flink connector](https://github.com/streamnative/pulsar-flink).
 {{< /hint >}}
```
[the remark this answers is not in the document]
Comment (imaffe) [reply docs/content/docs/connectors/datastream/pulsar.md]: Removed ~

File: docs/content/docs/connectors/datastream/pulsar.md (hunk 15 above)
Comment (imaffe) [reply docs/content/docs/connectors/datastream/pulsar.md:769]: Added~

**Comment by syhily** [comment]
@flinkbot run azure

**Comment by imaffe** [comment]
@MartijnVisser  Hi Martin, the CI has succeed ~

**Comment by MartijnVisser** [comment]
@imaffe Thanks, I've merged it. Will you also create the backport to the `release-1.15` branch?

**Comment by syhily** [comment]
@MartijnVisser OK, submitted. https://github.com/apache/flink/pull/19267

**Pull request merged**